### ISSUE TITLE
Add validation data loaders and logic

### ISSUE THREAD

**Issue by g8a9** [MELT-proj/training#18]
The current codebase does not prepare validation data loaders when the user specifies any in the config or via the CLI. We will need that to monitor both training and validation loss. The right place to add this is in the `trainer.py` module.

# Loss computation

This part should be straightforward. Moreover, we might want to support the option to select a maximum number of segments to be included in the evaluation split. Probably we need a more general feature (applicable to training CutSets, too) that lets you **subsample** a short cutset, maybe randomly, but still stratified by length (we don't want to validate only on audio of similar length).

# Generation-based, Extrinsic Metrics with Seq2SeqTrainer

My intuition is that we'll monitor both validation loss and extrinsic metrics computed on the generated text (when training for SFT/DPO) e.g., WER on translations. I.e., we might want to switch to a Seq2SeqTrainer because of this (here an excerpt of differences generate by copilot also by inspecting how ultravox uses it):

## 1) Trainer vs Seq2SeqTrainer: the main conceptual difference

Both classes run the same *training loop* (optimizer step, gradient accumulation, logging, checkpointing, distributed training, etc.). The key difference is what they do during **evaluation / prediction** (i.e., `evaluate()` / `predict()`), and therefore what kind of “predictions” they return for metrics.

### `Trainer` (generic)
- Assumes your model’s forward pass returns **logits** (e.g., next-token logits or class logits).
- In `prediction_step`, it typically:
  1. Runs `outputs = model(**inputs)` (no `generate`)
  2. Computes loss (if labels exist)
  3. Returns `(loss, logits, labels)`
- If you compute metrics, you usually compute them from **logits** (or argmax of logits).

This works well for classification, token classification, and even language modeling *if your metrics are token-level and based on teacher-forced logits*.

### `Seq2SeqTrainer` (generation-aware)
`Seq2SeqTrainer` is a subclass of `Trainer` specialized for tasks where you care about **generated sequences** (translation, summarization, “respond with text”, etc.).

Its main additions are:

1. **`predict_with_generate` support** via `Seq2SeqTrainingArguments`  
   In Transformers, `Seq2SeqTrainingArguments` adds:
   - `predict_with_generate` (bool)
   - `generation_max_length`, `generation_num_beams`, `generation_config`, etc.  
   Source:  
   https://github.com/huggingface/transformers/blob/5c77298c9379df31c4a2c859688197f3ac71d125/src/transformers/training_args_seq2seq.py#L44-L63

2. **Overrides `prediction_step` to call `model.generate()` (when enabled)**  
   In `Seq2SeqTrainer.prediction_step`, if `self.args.predict_with_generate` is true, it:
   - prepares “generation inputs”
   - calls `generated_tokens = self.model.generate(**generation_inputs, **gen_kwargs)`
   - pads outputs to max length
   - still optionally computes the loss with a normal forward pass
   - returns `(loss, generated_tokens, labels)` instead of `(loss, logits, labels)`  
   Source:  
   https://github.com/huggingface/transformers/blob/5c77298c9379df31c4a2c859688197f3ac71d125/src/transformers/trainer_seq2seq.py#L284-L370

3. **Generation config plumbing** (so eval/predict uses the right decoding settings)  
   `Seq2SeqTrainer` will also override/initialize the model’s `generation_config` from args if provided, and it carries `gen_kwargs` through `evaluate()`/`predict()`.  
   Sources:  
   - init / generation_config override: https://github.com/huggingface/transformers/blob/5c77298c9379df31c4a2c859688197f3ac71d125/src/transformers/trainer_seq2seq.py#L70-L91  
   - `evaluate()` setting `_gen_kwargs`: https://github.com/huggingface/transformers/blob/5c77298c9379df31c4a2c859688197f3ac71d125/src/transformers/trainer_seq2seq.py#L173-L191

### What this means in practice
- **Trainer** evaluation answers: “What logits does the model assign to the reference tokens?” (teacher forcing)
- **Seq2SeqTrainer** evaluation answers: “What text does the model actually generate under my decoding policy?” (beam search / sampling)

So yes: it is directly related to using `.generate()`—but specifically *during eval/predict so that metrics reflect real decoding behavior*, not teacher-forced logits.

---

## 2) Why Seq2SeqTrainer matters even for “causal LM chat” style models

A common confusion: “Seq2SeqTrainer is for encoder-decoder models (T5/BART), why use it for an LLaMA-style causal LM?”

Because “seq2seq” here is really shorthand for **sequence generation evaluation**, not strictly “encoder-decoder architecture”.

Even with a decoder-only model, if your task is “produce a response string”, your *quality metrics* (WER, ROUGE, exact match, task success, etc.) often need **generated text**, not logits.

`Trainer` can only easily give you logits; you could decode those logits greedily per time step, but:
- that’s not the same as `generate()` with KV caching, stopping criteria, min/max tokens, repetition penalty, etc.
- it won’t match inference behavior if you use beam search/sampling at inference time

`Seq2SeqTrainer` solves this by making evaluation/prediction “generation-native”.

---

## 3) How Ultravox uses Seq2SeqTrainer (and why)

Ultravox defines:

```python
class UltraTrainer(transformers.Seq2SeqTrainer):
    ...
```

Source:  
https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/training/trainer.py#L1-L8

Then in its training entrypoint it constructs the trainer with **`Seq2SeqTrainingArguments`**:

```python
trainer = ultra_trainer.UltraTrainer(
    ...
    args=transformers.Seq2SeqTrainingArguments(...),
)
```

Source:  
https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/training/train.py#L209-L292

### Key point: Ultravox is a generative “respond with text” model
Ultravox’s model class is explicitly generation-capable:

- `UltravoxModel(..., GenerationMixin)`
- and it implements `generate(...)` that ultimately delegates to the underlying language model’s `generate` after preparing multimodal embeddings (text+audio).  

Source (generate delegates):  
https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/model/ultravox_model.py#L409-L426

So Ultravox’s *real output* at inference time is what `model.generate()` produces given audio+prompt.

### Why `Seq2SeqTrainer` is the natural fit
Using `Seq2SeqTrainer` enables Ultravox to:
- run evaluation that can (if configured) call `UltravoxModel.generate()` on validation batches
- compute metrics on the generated token sequences (or decoded strings) that reflect deployment behavior

If they used plain `Trainer`, evaluation would only give teacher-forced logits; they’d have to bolt on custom generation logic to get realistic metrics.

### Ultravox’s customizations
Ultravox subclasses `Seq2SeqTrainer` mainly to customize dataloaders (e.g., caching eval dataloaders / behavior around iterable datasets and accelerate preparation):

Source:  
https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/training/trainer.py#L9-L84  
(and continuing around eval loader caching):  
https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/training/trainer.py#L97-L97

So: Ultravox isn’t reinventing generation evaluation; it’s inheriting Transformers’ standard generation-aware evaluation behavior and only tweaking the input pipeline aspects it needs.

---

## 4) Clearing up the inference vs evaluation confusion

You wrote:

> “It seems related to the possibility of using a model’s `.generate()` method during inference…”

Important nuance:

- **Inference**: you can always call `model.generate()` yourself, regardless of `Trainer` vs `Seq2SeqTrainer`.
- **What Seq2SeqTrainer changes** is: *when you run `trainer.evaluate()` or `trainer.predict()`*, it can use `generate()` internally (if `predict_with_generate=True`) so that predictions/metrics match real inference.

So the choice matters mostly for:
- validation metrics during training
- saving “generated predictions” during `predict()`
- tasks where you select best checkpoint based on generation quality (not loss)

---

## 5) One more practical detail: inputs to `generate()` in Ultravox

Ultravox has multimodal inputs (audio fields) and a special collator:

- `DataCollatorForSeq2SeqWithAudio(transformers.DataCollatorForSeq2Seq)`  
  Source: https://github.com/fixie-ai/ultravox/blob/69ddc63b2d72be5e9a86f818315da72cec55a876/ultravox/model/ultravox_processing.py#L1-L387

And `UltravoxModel.generate()` accepts extra args like `audio_values`, `audio_token_start_idx`, etc. It merges audio embeddings into the prompt embeddings before delegating to the LLM’s generation.  
That’s exactly the kind of situation where having the trainer’s eval loop call the model’s real `generate()` is valuable: you validate the full multimodal generation pathway, not just text-only forward logits.

---

## 6) About code-search completeness
The code-search tool returns at most 10 results per search, so the Ultravox references above may be incomplete.

You can browse more matches in the GitHub UI here:
- Ultravox search: https://github.com/fixie-ai/ultravox/search?q=Seq2SeqTrainer&type=code
- Ultravox search (predict_with_generate): https://github.com/fixie-ai/ultravox/search?q=predict_with_generate&type=code
- Transformers search (Seq2SeqTrainer): https://github.com/huggingface/transformers/search?q=class+Seq2SeqTrainer&type=code

---

If you tell me what metric(s) Ultravox uses for validation (or point me to the config you’re running), I can connect the dots all the way through: where `predict_with_generate` is set, what generation kwargs are passed (max tokens, beams), and how the generated sequences are turned into metrics.
